sample/src/main/java/com/shopify/sample/view/cart/CartActivity.java

```diff
@@ -224,6 +235,33 @@ private void onWebCheckoutConfirmation(final Checkout checkout) {
     customTabsIntent.launchUrl(this, Uri.parse(checkout.webUrl));
   }
 
+  private void onShopPayCheckoutConfirmation(final Checkout checkout) {
+    Uri shopPayUri = buildShopPayURI(checkout);
+    CustomTabsIntent customTabsIntent = new CustomTabsIntent.Builder().build();
+    customTabsIntent.launchUrl(this, shopPayUri);
+  }
+
+  private Uri buildShopPayURI(Checkout checkout) {
+    String storeAuthority = Uri.parse(checkout.webUrl).getAuthority();
+    Uri.Builder shopPayBuilder = new Uri.Builder()
+            .scheme("https")
+            .authority(storeAuthority)
+            .appendPath("cart")
+            .appendEncodedPath(getVariantSlug(checkout))
+            .appendQueryParameter("payment", "shop_pay");
+    return shopPayBuilder.build();
+  }
+
+  private String getVariantSlug(Checkout checkout) {
+    return checkout.lineItems.stream().map(item -> getDecodedVariantId(item.variantId) + ":" + item.quantity).collect(Collectors.joining(","));
+  }
+
+  private String getDecodedVariantId(String variantId) {
+    String fullVariantId = new String(Base64.decode(variantId, Base64.DEFAULT), StandardCharsets.UTF_8);
+    List<String> elements = Arrays.asList(fullVariantId.split("/"));
+    return elements.get(elements.size() - 1);
+  }
+
```

Comment:
Same suggestion re: unit-testability and treating them as independent routines (i.e. not necessarily in the activity)

Reply:
Moved to a util file and added tests.